File: algorithms/[518]零钱兑换 II/solution.ts
// topics = ["动态规划"]

function change(amount: number, coins: number[]): number {
  // dp[i] 表示组成金额 i 使用的硬币组合数
  const dp = new Array<number>(amount + 1).fill(0);
  dp[0] = 1;

  // 这里是求组合数，必须注意遍历顺序
  // 外层遍历物品，内层遍历背包容量
  for (const coin of coins) {
    for (let i = 1; i < amount + 1; i += 1) {
      dp[i] += dp[i - coin] ?? 0;
    }
  }

  return dp[amount];
}

File: algorithms/[518]零钱兑换 II/solution.py
# topics = ["动态规划"]

from typing import List


class Solution:
    def change(self, amount: int, coins: List[int]) -> int:
        """
        完全背包 & 恰好装满，求方案数（组合数）
        """
        # dp[i] 表示组成金额 i 使用的硬币方案数
        dp = [0] * (amount + 1)
        dp[0] = 1

        # 先遍历物品， 再遍历背包大小！
        for coin in coins:
            for i in range(1, amount + 1):
                if i >= coin:
                    dp[i] += dp[i - coin]

        return dp[amount]
